### PR TITLE
feat(examples): add grok/lenet_emnist with AdamW, subset training, and multi-metric checkpointing

### PR DESCRIPTION
**Pull request by mvillmow**
## Summary

New `examples/grok/lenet_emnist/` — a grokking-experiment scaffold for LeNet-5 on a subsampled EMNIST training set.

Built in response to a user training the standard `examples/lenet_emnist/` for 500 epochs hoping to observe grokking (Power et al. 2022) and observing textbook overfitting instead (peak test acc 86.72% @ ep 246, regressed to 85.91% @ ep 500 while training loss kept falling). Two skills were captured in [ProjectMnemosyne](https://github.com/HomericIntelligence/ProjectMnemosyne) documenting why (`training-diagnosis-loss-accuracy-decoupling-overfitting`) and what to do instead (`training-grokking-preconditions-and-vision-recipe`). This PR implements the latter.

## What's new

- **`examples/grok/lenet_emnist/`** — fork of `lenet_emnist/` with:
  - AdamW (decoupled weight decay) replacing SGD, used via the bare functional `adamw_step()`. Manual gradient flow retained because the autograd substrate isn't yet ready for convnets (see #5452).
  - `--subset-size N` flag to slice the train set (grokking requires a memorizable subset).
  - Per-epoch structured log line: `EPOCH N train_loss= ... train_acc= ... test_loss= ... test_acc= ... weight_l2= ...` (regex-parseable by the analyzer).
  - `MultiMetricCheckpointer` with `--track-metric NAME:MODE,NAME:MODE,...` supporting `max`/`min`/`both` modes; the `both` mode additionally tracks `min_after_max` and `max_after_min` so post-peak transitions (the grokking signal) aren't lost to trivial epoch-1 minima.
- **`train.sh`** — single-file dispatcher with three execution profiles:
  - `dry-run` (~30s): 1 batch of 1 epoch on a 64-sample subset
  - `smoke` (~5–15min): verify Phase 1 (train_acc -> 100%) is reachable on a 1k subset
  - `full` (~hours/day): real grokking attempt — 30k epochs, log every 50, ckpt every 500
- **`analyze_phases.py`** — post-hoc log analyzer that detects Phase 1 completion (train_acc ≥ 99% for ≥3 consecutive logged epochs), Phase 3 grokking onset (test_acc jumps ≥20pp within a 200-epoch window while train_acc stays ≥99%), and the loss-vs-accuracy decoupling fingerprint of late-stage overfitting.
- **`README.md`** — theory, realistic expectations (LeNet+EMNIST grokking is *unlikely* — convs encode a generalizing prior; calibrated probability table included), CLI surface, and links to the two source skills + the four related optimizer issues.

## Verification

- ✅ `train.sh dry-run` exits 0 in ~30s
- ✅ Single EPOCH log line emitted with all 5 metrics
- ✅ 7 checkpoint subdirectories created (`test_acc_best`, `test_loss_best`, `test_loss_min`, `test_loss_min_after_max`, `test_loss_max_after_min`, `train_loss_best`, `weight_l2_norm_best`) + `final/`, each with weights and a `meta.json` sidecar
- ✅ `analyze_phases.py --mode dry-run` parses the log cleanly and reports the 5 metrics
- ✅ `pixi run pre-commit run --files ...` clean

## Known caveats (filed/documented, not blocking)

- `weight_l2= 0.0` in dry-run output — silent-catch upstream bug in `compute_tensor_l2_norm`. Doesn't crash; analyzer NaN-guards it. Pre-existing.
- `--track-metric` accepts one comma-separated string rather than repeated flags (the project's `create_training_parser()` doesn't yet support list-typed args). Documented in CLI table.
- `examples/lenet_emnist/` (the SGD baseline) is **not** modified — intentional per scope.

## Related issues (filed alongside this PR)

- #5452 — `feat(autograd): complete autograd substrate for convnet training` (prerequisite for porting examples off manual gradients)
- #5449 — `feat(optim): add Muon optimizer`
- #5450 — `feat(optim): add NorMuon optimizer`
- #5451 — `feat(optim): add Lion and Shampoo`

Once #5452 lands, this example will be the natural first consumer of the autograd path. Once #5449–#5451 land, it'll be the natural benchmark for optimizer comparisons on a controlled small-data regime.

## Honest expectation about grokking

This scaffold creates the *necessary* conditions for grokking (small subset, AdamW, strong weight decay, long horizon, full instrumentation). It does **not** guarantee grokking will be observed — that depends on whether LeNet's conv inductive bias allows a memorize-then-grok dynamic at all on natural-image data, which the published literature suggests is unlikely. The README's "Realistic Expectations" section calibrates this honestly. The scaffold's value is to enable the experiment cleanly, not to manufacture a result.

## Test plan

- [ ] CI green
- [ ] `train.sh dry-run` runs end-to-end (verified locally before push)
- [ ] No regression in existing `examples/lenet_emnist/` (untouched)
- [ ] Markdown lint clean, pre-commit clean (verified locally before push)

Co-Authored-By: Claude Opus 4.7 (1M context) <noreply@anthropic.com>

### CODE REVIEW

**Comment by mvillmow** [comment]
# Implementation Plan

I have enough grounding. The work shipped in commit f874d76a (PR #5453 merged 2026-05-26). The plan I output describes the implementation strategy that was used, with real file:line references.

# Implementation Plan

## Objective

Add `examples/grok/lenet_emnist/` — a grokking-experiment scaffold for LeNet-5 on a subsampled EMNIST training set. Forks the existing `examples/lenet_emnist/` and swaps SGD for AdamW, adds `--subset-size N` for memorizable training sets, emits a regex-parseable per-epoch metrics line, introduces a `MultiMetricCheckpointer` with `max`/`min`/`both` modes (the latter tracking `min_after_max` and `max_after_min` post-peak transitions), ships a three-profile `train.sh` dispatcher (`dry-run` / `smoke` / `full`), a post-hoc `analyze_phases.py` phase detector, and a `README.md` calibrating expectations. Built in response to a user seeing textbook overfitting (peak test acc 86.72% @ ep 246, regressing to 85.91% @ ep 500) when training the standard SGD example for 500 epochs hoping to observe grokking.

## Approach

**Fork, don't extend.** Verified `examples/lenet_emnist/` is the canonical SGD baseline (`grep -n 'sgd_step\|SGD' examples/lenet_emnist/train.mojo` confirms manual SGD-style updates; no AdamW path). Leaving the original untouched preserves the per-PR baseline expectations called out in `agents/hierarchy.md` and the issue's scope ("`examples/lenet_emnist/` is **not** modified — intentional per scope"). Grokking changes are surgical enough (optimizer swap + subset slice + logging format + checkpointer + harness) that a parallel directory is cleaner than a flag matrix on the original.

**Bare functional `adamw_step()`, not the autograd path.** `examples/lenet_emnist/train_autograd.mojo` already exists and the autograd substrate is incomplete for convnets (tracked in #5452). Using `projectodyssey.training.optimizers.adamw.adamw_step` keeps manual gradient flow consistent with the SGD baseline at `examples/lenet_emnist/train.mojo`; once #5452 lands this example becomes the natural first consumer of the autograd path.

**Single regex-parseable EPOCH line.** Grepped existing log formats — none are stable for post-hoc parsing across runs. The grokking analyzer needs one anchor (`EPOCH N train_loss= ... train_acc= ... test_loss= ... test_acc= ... weight_l2= ...`) per logged epoch so an interrupted run still parses cleanly. The pattern is documented in `analyze_phases.py:46` and consumed by `EPOCH_RE = re.compile(r"EPOCH\s+(?P<epoch>\d+)(?P<rest>.*)")` at `analyze_phases.py:49`, with permissive key=value extraction so adding metrics later doesn't break readers.

**`both` mode for grokking fingerprinting.** A naïve `min`-mode test_loss checkpointer captures the trivial epoch-1 minimum and discards the late-stage Phase 3 transition. The `MultiMetricCheckpointer` defined at `run_train.mojo:342` resolves this by emitting four sub-checkpoints for a `both` spec: `<metric>_min`, `<metric>_max`, `<metric>_min_after_max`, `<metric>_max_after_min`, with each tracked independently against running peak state. Aligns with the team-knowledge `spec-driven-experimentation` and `defensive-analysis-patterns` skills called out in Prior Learnings.

**Single comma-separated `--track-metric` string.** `projectodyssey.training.cli.create_training_parser` does not yet support repeated/list-typed flags (verified by reading the parser; only scalar `get_string`/`get_int` accessors exist). Parsing `"test_acc:max,test_loss:both"` inside `parse_track_metric()` at `run_train.mojo:290` is the right scope for this PR; upgrading the parser is out of scope.

**Three-profile harness.** `dry-run` (~30 s) exists to validate the full checkpoint + log pipeline before burning compute on smoke or full. Matches the e2e-experiment-run-triage-completion skill's controlled-concurrency pattern: each profile is independently runnable and produces the same artifact shape (log + checkpoints + meta.json).

## Files to Create

### `examples/grok/lenet_emnist/model.mojo`
Fork of `examples/lenet_emnist/model.mojo`. Module path advertised as `grok/lenet_emnist/model`. Identical LeNet-5 architecture (2 conv + 2 FC + log_softmax) — the grokking change is in the training recipe, not the model.

### `examples/grok/lenet_emnist/train.mojo`
Manual forward + backward closure used by `run_train.mojo`. Returns the per-batch gradient bundle consumed by `adamw_step()`. Fork of `examples/lenet_emnist/train.mojo` with SGD removed (optimizer step moves to the driver).

### `examples/grok/lenet_emnist/run_train.mojo`
Driver. Sections:
- `TrainConfig` struct with `subset_size`, `track_metric`, `weight_decay`, `learning_rate`, `betas`, etc.
- `compute_weight_l2_norm(model)` at `run_train.mojo:262` for the `weight_l2=` metric.
- `parse_track_metric(spec)` at `run_train.mojo:290` — parses `"a:max,b:min,c:both"` into a `List[MetricMode]`.
- `struct MultiMetricCheckpointer` at `run_train.mojo:342` — owns one `(running, current)` pair per sub-metric and writes weights + `meta.json` per checkpoint subdirectory.
- Subset slice at `run_train.mojo:581–583`:
  ```mojo
  if config.subset_size > 0 and config.subset_size < train_images.shape()[0]:
      train_images = train_images.slice(0, config.subset_size, axis=0)
      train_labels = train_labels.slice(0, config.subset_size, axis=0)
  ```
- Per-epoch log line at `run_train.mojo:643` printing the five metrics in the canonical order so `analyze_phases.py` regex-matches.
- AdamW step invocation per layer: `adamw_step(param, grad, m, v, step, lr, beta1, beta2, eps, weight_decay)`.

### `examples/grok/lenet_emnist/run_infer.mojo`
Loader-side wrapper that consumes any of the seven checkpoint subdirectories (e.g. `test_acc_best/`, `test_loss_min_after_max/`) so the user can run inference against the grokking-onset weights specifically.

### `examples/grok/lenet_emnist/train.sh`
84-line dispatcher. Cases at lines 37 / 46 / 54:
```bash
dry-run) extra="--subset-size 64 --num-epochs 1 --log-every 1 --batch-size 64";;
smoke)   extra="--subset-size 1000 --num-epochs 50 --log-every 5";;
full)    extra="--subset-size 30000 --num-epochs 30000 --log-every 50 --checkpoint-every 500";;
```
Default `--track-metric` for `full` is `"test_acc:max,test_loss:both,train_loss:min,weight_l2_norm:min"`.

### `examples/grok/lenet_emnist/analyze_phases.py`
Post-hoc log analyzer. Pieces:
- `EPOCH_RE` at line 49.
- `parse_log(path)` at line 59 — tolerates non-EPOCH chatter (interrupted runs, framework banners).
- `detect_nans(rows)` at line 79 — guards against the known `weight_l2 = 0.0` silent-catch upstream bug.
- `diagnose_overfitting(rows)` at line 88 — loss-vs-accuracy decoupling fingerprint (training_loss falls ≥10 logged epochs while test_acc plateaus/regresses).
- `detect_phase1(rows, threshold=99.0, consecutive=3)` at line 125.
- `detect_phase3(rows, jump_pp=20.0, window=200)` at line 154.
- `emit_dry_run(rows)` at line 182 — `--mode dry-run` exit-0 verifier (≥1 EPOCH line, NaN-free).

### `examples/grok/lenet_emnist/README.md`
Theory (Power et al. 2022 grokking preconditions), calibrated-probability expectations table (LeNet's conv prior is *generalizing* — published evidence for grokking on natural-image conv nets is thin), the CLI surface table (documenting the single-string `--track-metric` limitation), and links to the two source skills and #5452 / #5449 / #5450 / #5451.

## Files to Modify

_None._ The fork-don't-extend decision is deliberate: `examples/lenet_emnist/` stays the SGD baseline.

## Implementation Order

1. **Create `examples/grok/` and `examples/grok/lenet_emnist/`.** Empty directories first so `pixi run mojo build` discovers the new package root.
2. **Fork `model.mojo`** verbatim from `examples/lenet_emnist/model.mojo`. Verify `pixi run mojo build examples/grok/lenet_emnist/model.mojo` succeeds before changing anything else.
3. **Fork `train.mojo`** — strip SGD-side updates so the per-batch step returns gradients to the driver. Compile-check.
4. **Write `run_train.mojo`** in this order so each layer compiles in isolation: (a) `TrainConfig`, (b) `compute_weight_l2_norm`, (c) `parse_track_metric` + `MetricMode`, (d) `MultiMetricCheckpointer`, (e) main loop with subset slice and the per-epoch log line, (f) AdamW invocation.
5. **Write `run_infer.mojo`** — copies the loader half of `examples/lenet_emnist/run_infer.mojo` and points it at a configurable checkpoint subdirectory.
6. **Write `analyze_phases.py`** test-first against synthetic logs (1 epoch row, NaN row, Phase 1 row sequence, Phase 3 jump sequence) before wiring it to real output.
7. **Write `train.sh`** — keep it short, fail loudly on bad mode, forward unknown flags to the Mojo driver (`"$@"`).
8. **Write `README.md`** last — theory + expectations + CLI table + links to skills/issues.
9. **Run dry-run end-to-end:** `./train.sh dry-run` and inspect `weights/*/meta.json` for all seven checkpoint subdirs.
10. **Pre-commit on the full diff** per memory `feedback_precommit_all_pr_files.md`: `pixi run pre-commit run --from-ref origin/main --to-ref HEAD`, plus `pixi run mojo format $(git diff --name-only origin/main HEAD -- '*.mojo')` per `feedback_precommit_mojo_format_diff_mode.md` (the `--from/--to-ref` form does not fire mojo-format).
11. **Open PR** with `gh pr create --body "Closes #5453"` and `gh pr merge --auto --rebase`.

## Verification

```bash
# Acceptance: train.sh dry-run exits 0 in ~30s
cd examples/grok/lenet_emnist && time ./train.sh dry-run; echo "exit=$?"

# Acceptance: single EPOCH log line with all 5 metrics is emitted
grep -E "^EPOCH 1 train_loss= .* train_acc= .* test_loss= .* test_acc= .* weight_l2= " \
  examples/grok/lenet_emnist/logs/dry-run.log

# Acceptance: 7 checkpoint subdirs + final/, each with weights and meta.json
ls examples/grok/lenet_emnist/weights/ | sort
test -f examples/grok/lenet_emnist/weights/test_acc_best/meta.json
test -f examples/grok/lenet_emnist/weights/test_loss_min_after_max/meta.json
test -f examples/grok/lenet_emnist/weights/test_loss_max_after_min/meta.json

# Acceptance: analyze_phases.py parses dry-run log cleanly
pixi run python examples/grok/lenet_emnist/analyze_phases.py \
  --mode dry-run --log examples/grok/lenet_emnist/logs/dry-run.log; echo "exit=$?"

# Acceptance: SGD baseline at examples/lenet_emnist/ is untouched
git diff --quiet origin/main -- examples/lenet_emnist/ && echo "baseline untouched"

# Acceptance: pre-commit clean on full PR diff (per feedback_precommit_all_pr_files)
pixi run pre-commit run --from-ref origin/main --to-ref HEAD
pixi run mojo format $(git diff --name-only origin/main HEAD -- '*.mojo')

# Acceptance: CI green
gh pr checks
```

## Skills Used

Planning-time tools: `Bash` (ls, grep, git, gh, wc, diff), `Read`-equivalent file inspection of the existing `examples/lenet_emnist/` baseline to ground the fork decision.

Team knowledge-base skills from Prior Learnings I applied:
- **spec-driven-experimentation** — three-profile harness (`dry-run` / `smoke` / `full`) with success thresholds (Phase 1 train_acc ≥99% × 3 consecutive logged epochs; Phase 3 test_acc jump ≥20 pp within 200-epoch window) defined before any expensive run.
- **defensive-analysis-patterns** — `analyze_phases.py` NaN-guards the `weight_l2 = 0.0` upstream silent-catch bug at `analyze_phases.py:79` rather than crashing.
- **e2e-experiment-run-triage-completion** — three independently-runnable profiles, each producing the same artifact shape (log + 7 ckpt subdirs + meta.json).
- **training-tests-weekly-workflow** — `full` profile is explicitly out-of-band from per-PR CI; the per-PR signal is `dry-run`.
- **optimization-mojo-simd-nan-inf-vectorization** — informed the `compute_weight_l2_norm` Float64 accumulation choice at `run_train.mojo:262` and the analyzer NaN-guard.
- **ci-cd-failure-diagnosis-log-analysis** — drove the permissive `EPOCH_RE` regex at `analyze_phases.py:49` (anchor on `EPOCH N`, then key=value extraction) so future metric additions don't break readers.

Memory entries applied:
- `feedback_precommit_all_pr_files` — verification uses `--from-ref origin/main --to-ref HEAD`, not staged-only.
- `feedback_precommit_mojo_format_diff_mode` — verification additionally runs `pixi run mojo format` on changed `.mojo` files because the diff-mode hook silently skips mojo-format.
- `feedback_no_shortcuts` — `train.sh dry-run` does not skip-if-exists; every run reproduces the artifact set from scratch.
- `normuon_optimizer_patterns` — referenced AdamW hyperparameter defaults (`lr=1e-3, weight_decay=1e-4, betas=(0.9, 0.999)`) consistent with the project's optimizer family.

## Changes from review

_N/A — initial plan._
